### PR TITLE
Add beam search peptide decoding

### PR DESCRIPTION
**Pull request by melihyilmaz**
Add beam search decoding to replace greedy decoding. Organized `beam_search_decode()` as a series of sub-functions to allow for unit testing each.

From a larger group of peptides predictions, this implementation of beam search decoding caches _k_ highest scoring peptide predictions for each spectrum, prioritizing the peptides fitting the observed precursor mass. As an output, the highest scoring peptide within precursor m/z tolerance, is returned, i.e. a single PSM for each spectra is recorded in the output mzTab file. If there are no cached peptides within precursor m/z tolerance for a spectrum, the highest scoring peptide  is returned in the output.

Also fixed amino acid-level score calculation in `model.on_predict_epoch_end()`, which was previously retrieving a list of shifted-by-one-AA amino acid scores, i.e. first AA prediction being assigned the second score, second AA the third score etc. where last AA was assigned the score for the stop toke.

### CODE REVIEW

File: casanovo/denovo/model.py [hunk 1]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                    pred_mz = self.peptide_mass_calculator.mass(
+                        aa_list, charge=precursors[beam_i, 1].item()
+                    )
+                    delta_mass_ppm = (pred_mz - prec_mz) / prec_mz * 10**6
+                    if delta_mass_ppm > -self.precursor_mass_tol:
+                        tokens[beam_i][idx] = self.stop_token
+
+        # Get the list of indices for finished beams
+        finished_beams_bool = (tokens == self.stop_token).any(dim=1)
+        finished_beams_idx = torch.where(finished_beams_bool == True)[
```
Comment (bittremieux) [inline casanovo/denovo/model.py:422]: **issue:** `== True` is not necessary.

File: casanovo/denovo/model.py [hunk 2]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+            0
+        ].tolist()
```
Comment (bittremieux) [inline casanovo/denovo/model.py:424]: **question:** Can we keep this as a tensor and avoid the memory copies (slow)?

File: casanovo/denovo/model.py [hunk 3]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+            Next available tensor index to cache peptides for each spectrum.
+        cache_seq_dict : Dict[int,  Set[str]]
+            Set of decoded peptides for each spectrum.
+        cache_score_dict : Dict[int, List[List[Tuple[float, int]]]
+            Confidence score for each decoded peptide, separated as precursor m/z fitting vs not, for each spectrum.
+
+        """
+        beam = self.n_beams
+        # Store finished beams in the cache
+        if len(finished_beams_idx) > 0:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:488]: **issue:** This `if` is not necessary; if the list is empty the iterator on the next line just won't run. Useful, because this method has a lot of indentations.

File: casanovo/denovo/model.py [hunk 4]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+            for i in finished_beams_idx:
+                spec_idx = i // beam  # find the starting index of the spectrum
+                insert_idx = cache_idx_dict[spec_idx]
+
+                # Only add beams with matching mass to the cache
+                prec_mz = precursors[i, 2].item()
+                aa_list = [
```
Comment (bittremieux) [inline casanovo/denovo/model.py:495]: **question:** We're decoding the AAs and calculating the mass both when terminating and caching beams. Can we avoid doing this twice?

File: casanovo/denovo/model.py [hunk 5]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                    self.decoder._idx2aa.get(idx_aa.item(), "")
+                    for idx_aa in tokens[i][:idx]
+                ]
+
+                # Check if dummy token predicted
+                if "" not in aa_list:
+                    # Remove stop token as needed, i.e. if mass filter induced termination
+                    if "$" in aa_list:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:503]: **suggestion:** Maybe the stop token and dummy token should be a member variable, rather than this "magic" string.

File: casanovo/denovo/model.py [hunk 6]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+
+                # Only add beams with matching mass to the cache
+                prec_mz = precursors[i, 2].item()
+                aa_list = [
+                    self.decoder._idx2aa.get(idx_aa.item(), "")
+                    for idx_aa in tokens[i][:idx]
+                ]
+
+                # Check if dummy token predicted
+                if "" not in aa_list:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:501]: **question:** In which situation is a predicted AA just empty?

File: casanovo/denovo/model.py [hunk 7]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                    # Remove stop token as needed, i.e. if mass filter induced termination
+                    if "$" in aa_list:
+                        aa_list.remove("$")
+
+                    pep_str = "".join(aa_list)
+                    pred_mz = self.peptide_mass_calculator.mass(
+                        aa_list, charge=precursors[i, 1].item()
+                    )
+                    delta_mass_ppm = (pred_mz - prec_mz) / prec_mz * 10**6
+                    isPrecursorFit = (
```
Comment (bittremieux) [inline casanovo/denovo/model.py:511]: **issue:** Use snake case as per the Python style conventions.

File: casanovo/denovo/model.py [hunk 8]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                        abs(delta_mass_ppm) <= self.precursor_mass_tol
```
Comment (bittremieux) [inline casanovo/denovo/model.py:512]: **issue:** Compare with `<` for consistent behavior with `on_predict_epoch_end`. These are floats anyway, so `<` and `<=` are semantically identical.

File: casanovo/denovo/model.py [hunk 9]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                            cache_scores[insert_idx, :, :] = scores[i, :, :]
+                            cache_idx_dict[spec_idx] += 1  # move the pointer
+                            cache_seq_dict[spec_idx].add(pep_str)
+                            smx = self.softmax(scores)[i, :, :]
+
+                            aa_scores = [
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            pep_score = np.mean(aa_scores)
```
Comment (bittremieux) [inline casanovo/denovo/model.py:529]: **suggestion:** It might be useful to extract this behavior to a AA-scores-to-peptide-scores function, so that if we want to change it, we can easily do it in a single location.

File: casanovo/denovo/model.py [hunk 10]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+
+                            # Cache peptides with fitting and non-fitting precursor mass separately
+                            if isPrecursorFit:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:532]: **suggestion:** Can be simplified as:
```
cache_i = not isPrecursorFit
heappush(cache_score_dict[spec_idx][cache_i], (pep_score, insert_idx))
```

File: casanovo/denovo/model.py [hunk 11]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                                    (pep_score, insert_idx),
+                                )
+                            else:
+                                heappush(
+                                    cache_score_dict[spec_idx][1],
+                                    (pep_score, insert_idx),
+                                )
+
+                        else:
+                            smx = self.softmax(scores)[i, :, :]
```
Comment (bittremieux) [inline casanovo/denovo/model.py:544]: **suggestion:** This behavior is duplicated in the `if` and the `else`, just do it once outside.

File: casanovo/denovo/model.py [hunk 12]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                            ]
+                            new_score = np.mean(aa_scores)
+
+                            # Cache fitting peptide either if non-fitting cached
+                            # or higher confidence than lowest scoring fitting peptide cached
+                            if isPrecursorFit:
+                                # Check if any non-fitting peptide cached
+                                if len(cache_score_dict[spec_idx][1]) > 0:
+                                    (
+                                        pop_pep_score,
```
Comment (bittremieux) [inline casanovo/denovo/model.py:557]: **todo:** No need to capture the results from the `heappop` because they're not used further.

File: casanovo/denovo/model.py [hunk 13]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                                        cache_score_dict[spec_idx][0],
+                                        (new_score, pop_insert_idx),
+                                    )
+
+                                # Check if any fitting peptide cached
+                                elif len(cache_score_dict[spec_idx][0]) > 0:
+                                    (
+                                        pop_pep_score,
+                                        pop_insert_idx,
+                                    ) = heappop(cache_score_dict[spec_idx][0])
```
Comment (bittremieux) [inline casanovo/denovo/model.py:576]: **suggestion:** Can we rewrite the succession of `heappush` and `heappop` more efficiently using `heappushpop`? Right now we're also popping and then pushing the same prediction back on the heap if the score was lower, which is a bit silly.

You can more nicely get the smallest element of a heap using `heap[0]`. That way you could "peek" at the value and only do a `heappushpop` if needed.

File: casanovo/denovo/model.py [hunk 14]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                                        heappush(
+                                            cache_score_dict[spec_idx][0],
+                                            (
+                                                pop_pep_score,
+                                                pop_insert_idx,
+                                            ),
+                                        )
+                            # Cache non-fitting peptide if higher confidence than lowest scoring non-fitting peptide cached
+                            else:
+                                if len(cache_score_dict[spec_idx][1]) > 0:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:598]: **suggestion:** Update the heap behavior similarly as above.

File: casanovo/denovo/model.py [hunk 15]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+            scores = einops.rearrange(scores, "B L V S -> (B S) L V")
+            tokens = einops.rearrange(tokens, "B L S -> (B S) L")
+
+            # Terminate beams exceeding precursor m/z tolerance and track all terminated beams
+            finished_beams_idx, tokens = self._terminate_finished_beams(
+                tokens, precursors, idx
+            )
+
+            # Cache terminated beams, group and order by fitting precursor m/z and confidence score
+            (
```
Comment (bittremieux) [inline casanovo/denovo/model.py:274]: **suggestion:** `cache_idx_dict`, `cache_seq_dict`, and `cache_score_dict` are passed by reference in the function, so you don't necessarily need to return them. I'd rewrite this function without a return value to simplify the code a bit.

File: casanovo/denovo/model.py [hunk 16]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+        output_scores = torch.zeros(batch, length, vocab)
+        output_scores = output_scores.type_as(cache_scores)
+        output_scores[output_scores == 0] = torch.nan
+        output_tokens = torch.zeros(batch, length)
+        output_tokens = output_tokens.type_as(cache_tokens)
+
+        # Return the top scoring peptide (fitting precursor mass if possible)
+        for spec_idx in range(batch):
+            cached_fitting, cached_nonfitting = cache_score_dict[spec_idx]
+            if len(cached_fitting) != 0:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:673]: **suggestion:** Can be slightly simplified as:
```
cache = cached_fitting if len(cached_fitting) > 0 else cached_nonfitting
_, top_score_idx = max(cache, key=lambda item: item[0])
```

File: casanovo/denovo/model.py [hunk 17]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                                smx[idx][self.decoder._aa2idx[aa]].item()
+                                for idx, aa in enumerate(aa_list)
+                            ]
+                            new_score = np.mean(aa_scores)
+
+                            # Cache fitting peptide either if non-fitting cached
+                            # or higher confidence than lowest scoring fitting peptide cached
+                            if isPrecursorFit:
+                                # Check if any non-fitting peptide cached
+                                if len(cache_score_dict[spec_idx][1]) > 0:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:555]: **todo:** Can we add a few unit tests that cover the below situations without test coverage?

File: casanovo/denovo/model.py [hunk 18]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+        ).any(axis=1)
+        # Mask out the index '0', i.e. padding token, by default
+        finished_mask[:, :beam] = True
+
+        # Figure out the top K decodings
+        _, top_idx = torch.topk(
+            step_scores.nanmean(dim=1) * (~finished_mask).int().float(),
+            beam,
+        )
+        V_idx, S_idx = np.unravel_index(top_idx.cpu(), (vocab, beam))
```
Comment (bittremieux) [inline casanovo/denovo/model.py:762]: **Issue:** Don't start variable names with a capital letter, as per the Python style convention.

File: casanovo/denovo/model.py [hunk 19]
```diff
@@ -411,11 +1008,14 @@ def on_predict_epoch_end(
...
-                    # Take the scores of the most probable amino acids.
-                    top_aa_scores = torch.max(
-                        aa_scores[1 : len(peptide_tokens) + 1], axis=1
-                    )[0]
-                    peptide_score = torch.mean(top_aa_scores).detach().item()
+                    # Take the scores corresponding to the predicted amino acids.
+                    top_aa_scores = [
+                        aa_scores[idx][self.decoder._aa2idx[aa]].item()
+                        for idx, aa in enumerate(
+                            list(reversed(peptide_tokens))
```
Comment (bittremieux) [inline casanovo/denovo/model.py:1015]: **issue:** No need for the `list` conversion.

File: tests/test_unit.py [hunk 20]
```diff
@@ -99,8 +100,163 @@ def request(self, *args, **kwargs):
...
 
     model = Spec2Pep()
     assert model.tb_summarywriter is None
+
+
+def test_beam_search_decode():
+    """
+    Test that beam search decoding and its sub-functions as intended during inference
+    """
+    model = Spec2Pep(n_beams=4)
```
Comment (bittremieux) [inline tests/test_unit.py:117]: **issue:** Can we have some small tests with different numbers of beams (1 should be there for sure).

**Review by bittremieux — CHANGES_REQUESTED** [review]
Nice work. This is non-trivial functionality to implement.

Here is some initial feedback. I'll probably want to do another thorough review after these comments have been addressed.

Some comments are relevant to multiple places in the code, but I haven't repeated the same thing multiple times, so be a bit mindful of that.

File: casanovo/config.yaml [hunk 21]
```diff
@@ -65,6 +65,7 @@ weight_decay: 1e-5
 # Training/inference options.
 train_batch_size: 32
 predict_batch_size: 1024
+n_beams: 1
```
[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/config.yaml:68]: `n_beams` determines both the number of beams explored and total number of beams cached. Model currently only reports a single PSM regardless of this parameter. We can add functionality to report multiple PSMs later.

I implemented beams explored = beams cached both because it's straight-forward but also because that's the minimum number of peptide predictions we are guaranteed to come up with for a given number of beams, i.e. larger cache than number of beamse doesn't guarantee more PSMs (if/when we start reporting multiple PSMs)

File: casanovo/denovo/model.py [hunk 22]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+                cache_tokens,
+                cache_scores,
+                tokens,
+                precursors,
+                scores,
+                idx,
+            )
+
+            # Stop decoding when all current beams are terminated
+            decoded = (tokens == self.stop_token).any(axis=1)
```
[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model.py:292]: Decoding stops when _all_ beams are terminated.

File: casanovo/config.yaml (hunk 21 above)
[the remark this answers is not in the document]
Comment (bittremieux) [reply casanovo/config.yaml:68]: Beams explored = beams cached is fine I think, but we will want an option (as a future feature addition is fine) to report multiple PSMs per spectrum.

When considering only a single beam, the result is the same as for greedy decoding. So we probably want to increase the default number of beams a bit to actually get advantage of this functionality.

[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/config.yaml:68]: Raising it to 5 and opening an issue for multi-PSM output for future reference

File: casanovo/denovo/model.py (hunk 4 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:495]: Will cache delta m/z to avoid duplication but couldn't come up with a straight forward way of caching the peptide to avoid idx2AA decoding twice i.e. beams can terminate when stop token predicted before precursor m/z is exceeded which causes all sorts of edge cases

File: casanovo/denovo/model.py [hunk 23]
```diff
@@ -176,12 +193,592 @@ def forward(
...
+            # Check only non-terminated beams
+            if self.stop_token not in tokens[beam_i]:
+                # Finish if dummy predicted at the previous step
+                if tokens[beam_i][idx - 1] == 0:
+                    tokens[beam_i][idx - 1] = self.stop_token
+                # Check delta m/z and terminate the beam if exceeds the threshold
+                else:
+                    prec_mz = precursors[beam_i, 2].item()
+                    aa_list = [
+                        self.decoder._idx2aa.get(idx_aa.item(), "")
```
[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Switched to `detokenize`

File: casanovo/denovo/model.py (hunk 6 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: It was a remnant from earlier debugging, now removed.

File: casanovo/denovo/model.py (hunk 9 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Extracted to `_aa_to_pep_score()`

File: casanovo/denovo/model.py (hunk 17 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Added more test cases to cover replacement of 1) non-fitting with fitting and 2) non-fitting with non-fitting  in the cache

File: tests/test_unit.py (hunk 20 above)
Comment (melihyilmaz) [reply tests/test_unit.py:134]: Added `n_beams=1` but couldn't think of other meaningful **small** tests to cover other beam sizes.

File: casanovo/denovo/model.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Regex logic in `peptide_mass_calculator.mass` can't handle cases where multiple N-terminal modifications are predicted (e.g. `-17.027+42.011`) if the peptide is `str` so it's useful to decode this way to have an AA list. Made this change elsewhere in `model.py` as well.

File: casanovo/denovo/model.py [hunk 24]
```diff
@@ -195,7 +196,18 @@ def forward(
             spectra.to(self.encoder.device),
             precursors.to(self.decoder.device),
         )
-        return [self.decoder.detokenize(t) for t in tokens], aa_scores
+        peptides = []
```
Comment (bittremieux) [inline casanovo/denovo/model.py:199]: **issue:** I don't understand this change. This copies the `detokenize` code from depthcharge and does exactly the same as what the previous one-line statement did, except that it's a list rather than a string.

1. Why is this change necessary? Both list and strings are iterables.
2. Even if this change is necessary, duplicating 95% of functionality from depthcharge is not the best solution.

File: casanovo/denovo/model.py [hunk 25]
```diff
@@ -418,8 +433,10 @@ def _terminate_finished_beams(
                 else:
                     precursor_mz = precursors[beam_i, 2].item()
                     precursor_charge = precursors[beam_i, 1].item()
-                    peptide_seq = self.decoder.detokenize(tokens[beam_i][:idx])
-
+                    peptide_seq = [
```
Comment (bittremieux) [inline casanovo/denovo/model.py:436]: **issue:** Similar question as above.

File: casanovo/denovo/model.py [hunk 26]
```diff
@@ -444,6 +461,37 @@ def _terminate_finished_beams(
                             abs(d) < self.precursor_mass_tol
                             for d in delta_mass_ppm
                         )
+
+                        # Don't terminate if m/z difference is by NH3 loss
```
Comment (bittremieux) [inline casanovo/denovo/model.py:465]: **issue:** This solution seems sub-optimal because it considers NH3 loss as a special token, whereas there are different N-terminal modifications as well. Why did NH3 loss get special treatment? I adds a decent amount of extra complexity and I don't understand why this is needed (not documented).

File: casanovo/denovo/model.py [hunk 27]
```diff
@@ -444,6 +461,37 @@ def _terminate_finished_beams(
...
+                        if (
+                            exceeds_precursor_mz_tol
+                            and not is_within_precursor_mz_tol
```
Comment (bittremieux) [inline casanovo/denovo/model.py:468]: **nitpick:** This is always the case if `exceeds_precursor_mz_tol` is True, hence not really needed.

File: casanovo/denovo/model.py [hunk 28]
```diff
@@ -444,6 +461,37 @@ def _terminate_finished_beams(
...
+                            and is_nh3_loss_in_vocab
+                        ):
+                            alt_peptide_seq = peptide_seq + ["-17.027"]
```
Comment (bittremieux) [inline casanovo/denovo/model.py:471]: **issue:** NH3 loss is an N-terminal modification, whereas here it's added at the end (i.e. C-terminal). So either the modification appears in the wrong location in the peptide string, or the peptide strings hasn't been reversed yet?
For mass calculation I guess the position doesn't really matter, but at the very least it's confusing. Worst case scenario it results in hard-to-diagnose bugs now or in the future.

File: casanovo/denovo/model.py [hunk 29]
```diff
@@ -444,6 +461,37 @@ def _terminate_finished_beams(
...
+                                for isotope in range(
+                                    self.isotope_error_range[0],
+                                    self.isotope_error_range[1] + 1,
+                                )
+                            ]
+                            is_precursor_mz_tol_w_loss = any(
+                                abs(d) < self.precursor_mass_tol
+                                for d in delta_mass_ppm
+                            )
+                            exceeds_precursor_mz_tol = (
```
Comment (bittremieux) [inline casanovo/denovo/model.py:491]: **nitpick:** Combine the previous statement with this one, i.e. `not any(...)`. Just doing `not` of another boolean is a bit silly.

**Review by bittremieux — CHANGES_REQUESTED** [review]
It's looking pretty good. I have a few comments / requests for clarification. I especially don't understand the latest NH3 loss-related changes.

File: casanovo/denovo/model.py [hunk 30]
```diff
@@ -171,16 +178,613 @@ def forward(
...
+        scores[:, :1, :, :] = einops.repeat(pred, "B L V -> B L V S", S=beam)
+
+        # Make precursors and memories the right shape for decoding:
+        precursors = einops.repeat(precursors, "B L -> (B S) L", S=beam)
+        mem_masks = einops.repeat(mem_masks, "B L -> (B S) L", S=beam)
+        memories = einops.repeat(memories, "B L V -> (B S) L V", S=beam)
+
+        # The main decoding loop:
+        for idx in range(1, self.max_length + 1):
+            scores = einops.rearrange(scores, "B L V S -> (B S) L V")
```
Comment (bittremieux) [inline casanovo/denovo/model.py:272]: **question:** Can this be moved outside of the loop?

File: casanovo/denovo/model.py (hunk 26 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:455]: NH3 loss is the only token with negative mass. In the previous implementation, early stopping kicked in the even if the beam was expecting a NH3 loss as the last token because m/z tolerance would be exceeded. I couldn't think of a more elegant way of covering this edge case and avoiding m/z threshold based stopping if the m/z margin is equivalent to NH3 loss

File: casanovo/denovo/model.py [hunk 31]
```diff
@@ -171,16 +178,613 @@ def forward(
...
+                tokens[~decoded, :idx],
+                precursors[~decoded, :],
+                memories[~decoded, :, :],
+                mem_masks[~decoded, :],
+            )
+
+            # Find top-k beams with highest scores and continue decoding those
+            scores, tokens = self._get_topk_beams(scores, tokens, batch, idx)
+
+        return self.softmax(output_scores), output_tokens
```
Comment (bittremieux) [inline casanovo/denovo/model.py:317]: **issue:** In the unrealistic situation that decoding of one or more beams didn't finish within `max_length` steps, `output_scores` and `output_tokens` will be undefined. Should we move the `return` up to within the `if decoded.all()` and appropriately handle the edge case when decoding didn't fully finish here?

File: casanovo/denovo/model.py [hunk 32]
```diff
@@ -171,16 +178,613 @@ def forward(
...
+        Returns
+        -------
+        finished_beams_idx : torch.Tensor
+            Indices of all finished beams on tokens tensor.
+        tokens : torch.Tensor of size (n_spectra * n_beams, max_length)
+            Output token of the model corresponding to amino acid sequences.
+        """
+        # Check if N-terminal NH3 loss is in the vocabulary
+        is_nh3_loss_in_vocab = (
+            self.peptide_mass_calculator.masses.get("-17.027", "") != ""
```
Comment (bittremieux) [inline casanovo/denovo/model.py:412]: **todo:** This can be simplified by `"-17.027" in self.peptide_mass_calculator.masses`.

File: casanovo/denovo/model.py [hunk 33]
```diff
@@ -171,16 +178,613 @@ def forward(
...
+        finished_mask = (
+            einops.rearrange(extended_prev_tokens, "B L V S -> B L (V S)")
+            == self.stop_token
+        ).any(axis=1)
+        # Mask out the index '0', i.e. padding token, by default
+        finished_mask[:, :beam] = True
+
+        # Figure out the top K decodings
+        _, top_idx = torch.topk(
+            step_scores.nanmean(dim=1) * (~finished_mask).int().float(), beam
```
Comment (bittremieux) [inline casanovo/denovo/model.py:772]: **question:** Weird to convert to int and then immediately to float.

File: casanovo/denovo/model.py (hunk 26 above)
Comment (bittremieux) [reply casanovo/denovo/model.py]: Ok, I didn't think/realize that NH3 loss is the only negative mass. Then it is indeed a bit of a special case, but we can try to generalize the code a little bit after the remaining comments have been addressed.

File: casanovo/denovo/model.py (hunk 24 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:190]: Regex logic in `peptide_mass_calculator.mass` downstream can't handle cases where multiple N-terminal modifications are predicted back to back (e.g. -17.027+42.011) if the peptide is `str` rather than list so it's useful to decode this way to have an AA list. We can implement a new regex to convert `str` to list here or `peptide_mass_calculator.mass` can be modified in DepthCharge .

File: casanovo/denovo/model.py [hunk 34]
```diff
@@ -176,12 +191,602 @@ def forward(
...
+                                precursor_mz,
+                                precursor_charge,
+                                isotope,
+                            )
+                            for isotope in range(
+                                self.isotope_error_range[0],
+                                self.isotope_error_range[1] + 1,
+                            )
+                        ]
+                        exceeds_precursor_mz_tol = any(
```
[the remark this answers is not in the document]
Comment (melihyilmaz) [reply casanovo/denovo/model.py:442]: I think this makes sense but should the same logic apply to `is_within_precursor_mz_tol` as well? It wouldn't have downstream implications for the beam as long as`exceeds_precursor_mz_tol = False` but it'd still lead to discrepancy between the two.

File: casanovo/denovo/model.py (hunk 27 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:454]: They can both be True at the same time (`exceeds_precursor_mz_tol` is True for any mass delta higher than lower bound - so it includes the tolerance range), in which case we wouldn't want to go check for NH3 loss.

File: casanovo/denovo/model.py (hunk 31 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:315]: How should we handle case if decoding didn't fully finish?

File: casanovo/denovo/model.py (hunk 30 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Moved it out of the loop and added reshaping inside `_get_topk_beams `

File: casanovo/denovo/model.py (hunk 28 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Peptide wasn't reversed at that point. Changed to this have peptide reversed before hand and modification added to N-terminus. Can revisit after `detokenize` resolved

File: casanovo/denovo/model.py (hunk 24 above)
Comment (bittremieux) [reply casanovo/denovo/model.py:190]: I prefer to keep the (de)tokenizing functionality fully on the depthcharge side, rather than mixing it in the different codebases. I suggest to revert the changes here and directly open a PR in depthcharge that will need to be integrated prior to merging this PR.

File: casanovo/denovo/model.py (hunk 31 above)
Comment (bittremieux) [reply casanovo/denovo/model.py:315]: My suggestion: if at least one beam finished, select from the finished beams. If no beam finished at all, just return dummy values (zero). We can write a unit test for this behavior by setting `max_length` to some ridiculously small number to verify that our implementation is correct, although this will likely never occur in reality (or at least it shouldn't).

Comment (melihyilmaz) [reply casanovo/denovo/model.py:318]: Made the changes you requested and implemented a test controlling if a finished beam is return if there are any. However, couldn't find a straightforward way to test if dummy values returned when `max_length` very small. Should I initialize a fixed model with known output? If not feel free to resolve the conversation

File: casanovo/denovo/model.py (hunk 24 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py]: Reverted to `detokenize` and updated depthcharge dependency

**Comment by melihyilmaz** [comment]
Added a small fix to negative mass-aware termination (to ensure we're not terminating a beam if doesn't exceed tolerance under _any_ negative mass token) and unit tests covering that functionality. Only that and 1 other unresolved conversation above remains.

File: casanovo/denovo/model.py (hunk 31 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:308]: @bittremieux in case you haven't seen my earlier comment

File: casanovo/denovo/model.py [hunk 35]
```diff
@@ -408,21 +875,27 @@ def on_predict_epoch_end(
...
-                        aa_scores[1 : len(peptide_tokens) + 1], axis=1
-                    )[0]
-                    peptide_score = torch.mean(top_aa_scores).detach().item()
+                for spectrum_i, precursor, aa_tokens, aa_scores in zip(*step):
+                    # Omit stop token.
+                    aa_tokens = aa_tokens[1:]
+                    peptide = "".join(aa_tokens)
+                    # If this is a non-finished beam (after exceeding
+                    # `max_length`), return a dummy (empty) peptide.
+                    if len(peptide) == 0:
```
Comment (bittremieux) [inline casanovo/denovo/model.py:884]: **question:** I don't understand these two lines. `peptide` is created from `aa_tokens` and only has length 0 if `aa_tokens` is empty, so this sets it from an empty list to an empty string? Both are empty iterables, so this code doesn't really do anything?

Comment (melihyilmaz) [reply casanovo/denovo/model.py:884]: If `aa_tokens` has a number of empty strings `len(aa_tokens)>0` which we try to iterate over below. The list of empty strings arise from dummy tokens.

Comment (bittremieux) [reply casanovo/denovo/model.py:884]: Got it, thanks for the explanation. Maybe we should add a comment with that info for future reference.

However, I think that the code will crash in some of the following lines in that case. For example, if `aa_tokens` is empty, `top_aa_scores` will also be an empty list, and taking the mean when calculating the peptide score will crash.

So I don't think we have a test for this situation yet? And maybe rather than setting `aa_scores` to empty, we should not do any of the following code and directly give a 0-score to the mzTab writer?

Comment (bittremieux) [reply casanovo/denovo/model.py:884]: Can you also verify that the AA scores and peptide sequence are correctly ordered? Right now we take the peptide from the `aa_tokens` directly in line 881, but we ultimately reverse the AA scores before writing them in line 925. It seems to me that those two operations are not in sync with each other.

Comment (melihyilmaz) [reply casanovo/denovo/model.py:884]: re: empty list, as `np.mean` returns NaN without crashing the method doesn't crash and outputs empty string for the peptide and NaN for score in the mzTab. I think this can stay as is but I can implement a form of zero output if you think that's better.

Comment (melihyilmaz) [reply casanovo/denovo/model.py]: re: AA scores and tokens order, I corrected the bug introduced in [053967e](https://github.com/Noble-Lab/casanovo/pull/87/commits/053967edc80b31c998dcb446c7aae861d04b3fd8), refactored the code and added tests for this functionality.

File: casanovo/denovo/model.py (hunk 31 above)
Comment (melihyilmaz) [reply casanovo/denovo/model.py:303]: I added a test case.

File: casanovo/denovo/model.py (hunk 35 above)
Comment (bittremieux) [reply casanovo/denovo/model.py]: Woopsie. Thanks for catching that!

I was also too fast in checking `np.mean`, which indeed doesn't crash on empty input but gives a `RuntimeWarning`. I saw a stack trace and didn't look closely enough. But this works then.

File: casanovo/denovo/model.py [hunk 36]
```diff
@@ -460,6 +928,52 @@ def on_predict_epoch_end(
...
+        aa_tokens : List[str]
+            Amino acid tokens of the peptide sequence.
+        peptide_score : str
+            Peptide-level confidence score.
+        aa_scores : str
+            Amino acid-level confidence scores for the predicted sequence.
+        """
+        # Omit stop token.
+        aa_tokens = aa_tokens[1:] if self.decoder.reverse else aa_tokens[:-1]
+        peptide = "".join(aa_tokens)
```
Comment (bittremieux) [inline casanovo/denovo/model.py:957]: **question:** One final confusion about the order of the tokens and AAs. Here we create `peptide` from `aa_tokens` without reversing them, whereas in line 969 the tokens are reversed prior to getting the AA scores. Doesn't that contradict itself?
Although I guess the scores are reversed twice in lines 969-970, but why do `aa_scores` only need the be reversed once?

Comment (melihyilmaz) [reply casanovo/denovo/model.py:957]: `aa_tokens` arrives in the correct order in any case (except for the position of stop token) but aa scores arrive reversed `if self.decoder.reverse = True` so that's why we a) reverse `aa_tokens` to correspond to the order of scores for retrieval and then b) retrieved aa scores are reversed to correspond to the N- to C-terminus order. Let me know if I'm missing anything.

Comment (bittremieux) [reply casanovo/denovo/model.py:957]: Ok, rather than reversing `aa_tokens` twice, wouldn't it be easier to reverse `aa_scores` only once? E.g.:
```
top_aa_scores = [
    aa_score[self.decoder._aa2idx[aa_token]].item()
    for aa_score, aa_token in zip(aa_scores[::step], aa_tokens)
]
```

Slightly less code, but also less confusing to understand what's happening I think.

Comment (melihyilmaz) [reply casanovo/denovo/model.py:957]: `aa_scores` is a tensor with size `(max_length, vocab)` so both the snippet you shared doesn't work as is and also reversal of the non-empty positions in the tensor wasn't as straightforward for me.

Comment (bittremieux) [reply casanovo/denovo/model.py:957]: That makes total sense! Thanks.

**Review by bittremieux — APPROVED** [review]
